### PR TITLE
Refactor PreTrainer and enhance validation and early stopping

### PR DESCRIPTION
**Pull request by aaishwarymishra**
This pull request refactors and improves the configuration handling, validation, and robustness of the training pipeline, especially around dataset splitting and early stopping. It also simplifies the `PreTrainer` class, adds more comprehensive tests, and enhances error handling for invalid configurations.

**Configuration validation and error handling:**

- Improved validation logic in `DataConfig` to ensure both `train_ratio` and `val_ratio` are strictly between 0 and 1, and added explicit error messages for invalid values.
- Added validation and error handling for an empty training dataset, raising a clear exception during loader construction.

**Trainer and early stopping enhancements:**

- Refactored the `PreTrainer` class to simplify its initialization, enforce required configuration, and move device/vocab/model/optimizer resolution into dedicated methods.
- Made early stopping patience configurable via `TrainerConfig`, with validation to ensure it is a positive integer or `None`. Early stopping is now disabled if no validation loader is provided. [[1]](diffhunk://#diff-73352dfb97a0b9afb3ad346b22af2fdb5df6e72c73da6aa0df96d85c0a4c0cddR28) [[2]](diffhunk://#diff-8f40d97481e59dd82f8477965646f5d9517f80b514ebf0d45e17346dae39c1caL367-R354)

**Testing improvements:**

- Expanded tests to cover more invalid `train_ratio` and `val_ratio` cases, and added tests for empty datasets and early stopping patience, ensuring robust error handling. [[1]](diffhunk://#diff-93b71e6c75994ca8837c83e65b07112874a099bae00af92f3dfc9f1e49ea3711L86-R110) [[2]](diffhunk://#diff-9293ca79a9504730d4ee5e2d3493e6f1b3629160dc0056275094240ab8836432L392-R427)
- Added an `EmptyDataset` class for use in tests, enabling explicit checks for empty dataset handling.

**Codebase simplification and cleanup:**

- Removed unused arguments and simplified the interface of `PreTrainer.set_loaders`, making loader setup more predictable and less error-prone.
- Cleaned up imports in `trainer.py` and related files to remove unused configuration imports and streamline dependencies. [[1]](diffhunk://#diff-db20bb2c084e91a3c3a59237abeb51782b470217c982dffc387b2860e3d218d1R6-R12) [[2]](diffhunk://#diff-db20bb2c084e91a3c3a59237abeb51782b470217c982dffc387b2860e3d218d1L20-L27) [[3]](diffhunk://#diff-8f40d97481e59dd82f8477965646f5d9517f80b514ebf0d45e17346dae39c1caL19-R19)

**Testing and validation consistency:**

- Adjusted and clarified test assertions for invalid ratio values, ensuring that both lower and upper bounds are checked for both `train_ratio` and `val_ratio`.

These changes collectively improve the reliability, maintainability, and clarity of the training pipeline and its configuration system.

Part of #35

### CODE REVIEW

File: tests/config/base_test.py [hunk 1]
```diff
@@ -83,7 +83,31 @@ def test_data_config_invalid_ratios():
...
+        ValidationError,
+        match="train_ratio must be between 0 and 1",
+    ):
+        DataConfig(dataset=ComponentConfig(_target_="dataset.all"), train_ratio=1.1)
+
+    with pytest.raises(
+        ValidationError,
+        match="val_ratio must be between 0 and 1",
+    ):
+        DataConfig(dataset=ComponentConfig(_target_="dataset.all"), val_ratio=0.0)
```
Comment (TahaZahid05) [inline tests/config/base_test.py]: val_ratio being 0.0 shouldn't raise an error?

File: tests/trainers/pretrainer_test.py [hunk 2]
```diff
@@ -389,13 +397,31 @@ def test_pretrainer_builds_train_val_and_test_loaders_from_ratios(tmp_path):
...
+        ValueError, match="early_stopping_patience must be a positive integer"
+    ):
+        trainer = PreTrainer(project_config)
+        trainer.run()
+    project_config.trainer.early_stopping_patience = 0
+    with pytest.raises(
+        ValueError, match="early_stopping_patience must be a positive integer"
+    ):
+        trainer = PreTrainer(project_config)
+        trainer.run()
```
Comment (TahaZahid05) [inline tests/trainers/pretrainer_test.py]: Fix this please. same test being repeated twice.

File: trainite/config/base.py [hunk 3]
```diff
@@ -80,24 +81,25 @@ def validate_options(self) -> "DataConfig":
...
             train_ratio = self.train_ratio if self.train_ratio is not None else 1.0
             val_ratio = self.val_ratio if self.val_ratio is not None else 0.0
 
-            if train_ratio <= 0.0 or val_ratio < 0:
+            if train_ratio <= 0.0 or train_ratio > 1.0:
                 raise ValueError(
                     f"train_ratio must be between 0 and 1. Got train_ratio={train_ratio}, val_ratio={val_ratio}"
                 )
 
+            if val_ratio <= 0.0 or val_ratio > 1.0:
```
Comment (TahaZahid05) [inline trainite/config/base.py]: If there is no val_split, it would be 0.0 and it won't be an error?

File: trainite/trainers/pretrainer.py [hunk 4]
```diff
@@ -162,45 +156,31 @@ def _build_loaders_from_ratios(
...
+            self.logger.info(
+                "Automatic splitting requested with ratios: train=%s, val=%s",
+                self.config.data.train_ratio,
+                self.config.data.val_ratio,
+            )
+            (
+                self.train_loader,
+                self.val_loader,
+                self.test_loader,
+            ) = self._build_loaders_from_ratios(self.config.data)
```
Comment (TahaZahid05) [inline trainite/trainers/pretrainer.py:177]: looks a bit confusing. maybe change it to:
```
self.train_loader, self.val_loader, self.test_loader = self._build_loaders_from_ratios(self.config.data)
```

File: trainite/trainers/pretrainer.py [hunk 5]
```diff
@@ -364,9 +344,14 @@ def score_function(engine):
         self.handlers["checkpoint_last"] = last_checkpoint
 
         # 4. EarlyStopping
-        if self.val_loader:
+        patience = self.config.trainer.early_stopping_patience
+        if self.val_loader and patience is not None:
+            if patience <= 0:
```
Comment (TahaZahid05) [inline trainite/trainers/pretrainer.py]: we can validate in pydantic early rather than here. let's move it, and also update test.

File: trainite/trainers/pretrainer.py (hunk 4 above)
Comment (aaishwarymishra) [reply trainite/trainers/pretrainer.py:177]: its ruff formatting

**Review by TahaZahid05 — APPROVED** [review]
